### PR TITLE
router: allowing request-header-based retry checks

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: Low (new code is config guarded)
Testing: new unit tests
Docs Changes: n/a
Release Notes: yes

### CODE REVIEW

File: test/common/router/retry_state_impl_test.cc [hunk 1]
```diff
@@ -642,6 +642,44 @@ TEST_F(RouterRetryStateImplTest, PolicyResetRemoteReset) {
...
+  matcher2->set_name(":method");
+  matcher2->set_exact_match("HEAD");
+
+  policy_.retriable_request_headers_ = Http::HeaderUtility::buildHeaderMatcherVector(matchers);
+
+  {
+    Http::TestHeaderMapImpl request_headers{{"x-envoy-retry-on", "5xx"}};
+    setup(request_headers);
+    EXPECT_FALSE(state_->enabled());
+    EXPECT_EQ(RetryStatus::No, state_->shouldRetryReset(overflow_reset_, callback_));
```
Comment (snowp) [inline test/common/router/retry_state_impl_test.cc]: probably not needed to check this?

File: source/common/router/config_impl.h [hunk 2]
```diff
@@ -260,6 +264,7 @@ class RetryPolicyImpl : public RetryPolicy {
   uint32_t host_selection_attempts_{1};
   std::vector<uint32_t> retriable_status_codes_;
   std::vector<Http::HeaderMatcherSharedPtr> retriable_headers_;
+  absl::optional<std::vector<Http::HeaderMatcherSharedPtr>> retriable_request_headers_;
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/router/config_impl.h]: Yeah, I assumed naively that it saved memory.  TIL.

I do worry about the memory impact of feature creep on the retry state, but given it's include not apis we can always switch to pointers (which definitely don't take as much space) at some point in the future, so I'll stick with the types we're already using for now.

File: api/envoy/api/v2/route/route.proto [hunk 3]
```diff
@@ -931,10 +931,13 @@ message RetryPolicy {
   // describes Envoy's back-off algorithm.
   RetryBackOff retry_back_off = 8;
 
-  // HTTP headers that trigger a retry if present in the response. A retry will be
+  // HTTP response headers that trigger a retry if present in the response. A retry will be
   // triggered if any of the header matches match the upstream response headers.
   // The field is only consulted if 'retriable-headers' retry policy is active.
   repeated HeaderMatcher retriable_headers = 9;
```
[the remark this answers is not in the document]
Comment (lizan) [reply api/envoy/api/v2/route/route.proto]: @htuch shall we handle field rename like this in `protoxform` when upgrade to next-major-version?

[the remark this answers is not in the document]
Comment (mattklein123) [reply api/envoy/api/v2/route/route.proto]: Yeah we should add this capability as an annotation or comment.

File: source/common/http/async_client_impl.h [hunk 4]
```diff
@@ -135,13 +135,17 @@ class AsyncStreamImpl : public AsyncClient::Stream,
...
+      return retriable_request_headers_;
+    }
     absl::optional<std::chrono::milliseconds> baseInterval() const override {
       return absl::nullopt;
     }
     absl::optional<std::chrono::milliseconds> maxInterval() const override { return absl::nullopt; }
 
     const std::vector<uint32_t> retriable_status_codes_{};
     const std::vector<Http::HeaderMatcherSharedPtr> retriable_headers_{};
+    const std::vector<Http::HeaderMatcherSharedPtr> retriable_request_headers_{};
```
Comment (mattklein123) [inline source/common/http/async_client_impl.h]: super nit: The `{}` is not needed on any of these.

File: api/envoy/api/v2/route/route.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/api/v2/route/route.proto]: Yes, I will be adding annotations for this later. For now, you can use `[#next-major-version:]` or file an explicit GitHub issue and put it in the v3 API label.

**Pull request merged**